uro98/assignment3/FindAllWords.java

```diff
@@ -0,0 +1,80 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class FindAllWords {
+  
+  /**
+   * Find all the words in the dictionary that can be formed in the grid.
+   * @param grid
+   * @param dictionary
+   * @return
+   */
+  public static Set<String> findAllWords(Grid grid, Dictionary dictionary) {
+    Set<String> totalWords = new HashSet<String>();
+    
+    // For each position in the grid, find all the words that can be formed starting from there
+    for (int i = 0; i < grid.getHeight(); i++) {
+      for (int j = 0; j < grid.getWidth(); j++) {
+        Set<String> wordSet = FindAllWords.searchWords(
+            grid,
+            dictionary,
+            new boolean[grid.getHeight()][grid.getWidth()],
+            new HashSet<String>(),
+            new ArrayList<String>(),
+            i,
+            j);
+        totalWords.addAll(wordSet);
+      }
+    }
+    
+    return totalWords;
+  }
+  
+  public static Set<String> searchWords(
+      Grid grid,
+      Dictionary dictionary,
+      boolean[][] visited,
+      Set<String> words,
+      List<String> path,
+      int posX,
+      int posY) {
+    
+    if (posX < 0 || posX >= grid.getHeight() || posY < 0 || posY >= grid.getWidth() || visited[posX][posY]) {
+      return new HashSet<String>();
+    }
+    
+    // Update to include new position
+    path.add(grid.getLetter(posX, posY));
+    visited[posX][posY] = true;
+    
+    String pathString = String.join("", path);
+    
+    if (!dictionary.isPrefix(pathString)) {
+      return new HashSet<String>();
+    }
+    
+    if (dictionary.isWord(pathString)) {
+      words.add(pathString);
+    }
+    
+    // Recursively search each of the 8 adjacent cells
+    for (int i = -1; i <= 1; i++) {
+      for (int j = -1; j <= 1; j++) {
+        // Make copies of path and visited to search for words in different states
+        List<String> pathCopy = new ArrayList<String>(path);
+        boolean[][] visitedCopy = new boolean[visited.length][];
+        for (int k = 0; k < visited.length; k++) {
+          visitedCopy[k] = visited[k].clone();
+        }
+        
+        Set<String> newWords = searchWords(grid, dictionary, visitedCopy, words, pathCopy, posX + i, posY + j);
```

Comment:
With this invocation, `newWords` does not actually contain new words. Depending on situation it contains one of:

- empty set if no new words have been found;
- set of all words found so far (both old and new), since you are passing the same `Set` object created in line 24 all the way down and modify it on the way.

There are a few options:

- Don't pass `words` as parameter at all and just create empty `Set` at the beginning of this method.
- Don't return final set from this method at all and just rely on the fact, that code will modify set that you passed here from line 24.
- Keep returning final set as you do now, since it is convenient in the place where you start recursion, but then you do not need `newWords` variable and no need to call `words.addAll(newWords)` below.

I think the first option is cleanest, but all of them are valid, so it is up to you which one to choose.

Reply:
Thank you for the detailed feedback. Just out of curiosity, why do you think the first option is the cleanest?